File: client/java/Ship.java
import com.sun.jna.Pointer;

///A space ship!
class Ship
{
  Pointer ptr;
  int ID;
  int iteration;
  public Ship(Pointer p)
  {
    ptr = p;
    ID = Client.INSTANCE.shipGetId(ptr);
    iteration = BaseAI.iteration;
  }
  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.ships.length; i++)
    {
      if(BaseAI.ships[i].ID == ID)
      {
        ptr = BaseAI.ships[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }
    
    //commands
    
  ///
  int move(int x, int y)
  {
    validify();
    return Client.INSTANCE.shipMove(ptr, x, y);
  }
  ///
  int selfDestruct()
  {
    validify();
    return Client.INSTANCE.shipSelfDestruct(ptr);
  }
  ///
  int attack(Ship target)
  {
    validify();
    target.validify();
    return Client.INSTANCE.shipAttack(ptr, target.ptr);
  }
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.shipGetId(ptr);
  }
  ///The owner of the piece
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.shipGetOwner(ptr);
  }
  ///Position x
  public int getX()
  {
    validify();
    return Client.INSTANCE.shipGetX(ptr);
  }
  ///Position y
  public int getY()
  {
    validify();
    return Client.INSTANCE.shipGetY(ptr);
  }
  ///ship size radius
  public int getRadius()
  {
    validify();
    return Client.INSTANCE.shipGetRadius(ptr);
  }
  ///The ship type
  public String getType()
  {
    validify();
    return Client.INSTANCE.shipGetType(ptr);
  }
  ///how many more attacks it has
  public int getAttacksLeft()
  {
    validify();
    return Client.INSTANCE.shipGetAttacksLeft(ptr);
  }
  ///how much more movement it has
  public int getMovementLeft()
  {
    validify();
    return Client.INSTANCE.shipGetMovementLeft(ptr);
  }
  ///the largest possible movement
  public int getMaxMovement()
  {
    validify();
    return Client.INSTANCE.shipGetMaxMovement(ptr);
  }
  ///the max number of attacks it has
  public int getMaxAttacks()
  {
    validify();
    return Client.INSTANCE.shipGetMaxAttacks(ptr);
  }
  ///the strength of its attacks
  public int getDamage()
  {
    validify();
    return Client.INSTANCE.shipGetDamage(ptr);
  }
  ///the total health of the ship
  public int getHealth()
  {
    validify();
    return Client.INSTANCE.shipGetHealth(ptr);
  }
  ///the max health possible for the ship
  public int getMaxHealth()
  {
    validify();
    return Client.INSTANCE.shipGetMaxHealth(ptr);
  }

}

File: client/csharp/Ship.cs
using System;
using System.Runtime.InteropServices;


///A space ship!
public class Ship: ShipDescription
{

  public Ship()
  {
  }

  public Ship(IntPtr p)
  {
    ptr = p;
    ID = Client.shipGetId(ptr);
    iteration = BaseAI.iteration;
  }

  public override bool validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.ships.Length; i++)
    {
      if(BaseAI.ships[i].ID == ID)
      {
        ptr = BaseAI.ships[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }

    //commands

  ///Command a ship to move to a specified position. If the position specified by this function is not legal, the position of the ship will be updated, but the movement will be rejected by the server.
  public int move(int x, int y)
  {
    validify();
    return Client.shipMove(ptr, x, y);
  }
  ///Blow yourself up, damage those around you, reduces the ship to 0 health.
  public int selfDestruct()
  {
    validify();
    return Client.shipSelfDestruct(ptr);
  }
  ///Commands your ship to attack a target. Making an attack will reduce the number of attacks available to the ship, even if the attack is rejected by the game server.
  public int attack(Ship target)
  {
    validify();
    target.validify();
    return Client.shipAttack(ptr, target.ptr);
  }

    //getters

  ///Unique Identifier
  public new int getId()
  {
    validify();
    int value = Client.shipGetId(ptr);
    return value;
  }
  ///The ship type
  public new string getType()
  {
    validify();
    IntPtr value = Client.shipGetType(ptr);
    return Marshal.PtrToStringAuto(value);
  }
  ///The amount of money required to purchase this type of ship
  public new int getCost()
  {
    validify();
    int value = Client.shipGetCost(ptr);
    return value;
  }
  ///The radius of the ship
  public new int getRadius()
  {
    validify();
    int value = Client.shipGetRadius(ptr);
    return value;
  }
  ///The range of attacks for this ship
  public new int getRange()
  {
    validify();
    int value = Client.shipGetRange(ptr);
    return value;
  }
  ///The strength of attacks for this ship
  public new int getDamage()
  {
    validify();
    int value = Client.shipGetDamage(ptr);
    return value;
  }
  ///The amount of damage done when this ship self destructs
  public new int getSelfDestructDamage()
  {
    validify();
    int value = Client.shipGetSelfDestructDamage(ptr);
    return value;
  }
  ///The largest possible movement for this ship
  public new int getMaxMovement()
  {
    validify();
    int value = Client.shipGetMaxMovement(ptr);
    return value;
  }
  ///The max number of attacks for this ship
  public new int getMaxAttacks()
  {
    validify();
    int value = Client.shipGetMaxAttacks(ptr);
    return value;
  }
  ///The max health possible for the ship
  public new int getMaxHealth()
  {
    validify();
    int value = Client.shipGetMaxHealth(ptr);
    return value;
  }
  ///The owner of the ship
  public int getOwner()
  {
    validify();
    int value = Client.shipGetOwner(ptr);
    return value;
  }
  ///X position of the ship
  public int getX()
  {
    validify();
    int value = Client.shipGetX(ptr);
    return value;
  }
  ///Y position of the ship
  public int getY()
  {
    validify();
    int value = Client.shipGetY(ptr);
    return value;
  }
  ///How many more attacks this ship has
  public int getAttacksLeft()
  {
    validify();
    int value = Client.shipGetAttacksLeft(ptr);
    return value;
  }
  ///How much more movement this ship has
  public int getMovementLeft()
  {
    validify();
    int value = Client.shipGetMovementLeft(ptr);
    return value;
  }
  ///The total health of the ship
  public int getHealth()
  {
    validify();
    int value = Client.shipGetHealth(ptr);
    return value;
  }

}
